sfms/sfms.py

```diff
@@ -0,0 +1,169 @@
+"""
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+
+This script is used to upload SFMS output to the Predictive Services Unit API.
+
+The code relating to multipart form data is based on the blog post:
+https://blog.thesparktree.com/the-unfortunately-long-story-dealing-with
+"""
+import sys
+import os
+import mimetools
+import mimetypes
+import io
+import httplib
+import urlparse
+import codecs
+
+
+def get_config(ini):
+    """ Some very simple code for reading a config file into a dictionary. """
+    file = None
+    try:
+        file = open(ini, 'r')
+        config = {}
+        for line in file.readlines():
+            line = line.strip()
+            if line.startswith('#'):
+                continue
+            if line:
+                key, value = line.split('=')
+                config[key] = value
+    finally:
+        if file:
+            file.close()
+    return config
+
+
+class MultiPartForm(object):
+    """ Accumulate the data to be used when posting a form.
+    Based on blogpost: https://blog.thesparktree.com/the-unfortunately-long-story-dealing-with    
+    """
+
+    def __init__(self):
+        self.form_fields = []
+        self.files = []
+        self.boundary = mimetools.choose_boundary()
+        return
+
+    def get_content_type(self):
+        return 'multipart/form-data; boundary=%s' % self.boundary
+
+    def add_field(self, name, value):
+        """Add a simple field to the form data."""
+        self.form_fields.append((name, value))
+        return
+
+    def add_file(self, fieldname, filename, mimetype=None):
+        """Add a file to be uploaded."""
+        fileHandle = None
+        try:
+            fileHandle = codecs.open(filename, "rb")
+            body = fileHandle.read()
+        finally:
+            if fileHandle:
+                fileHandle.close()
+        basename = os.path.basename(filename)
+        if mimetype is None:
+            mimetype = mimetypes.guess_type(basename)[0] or 'application/octet-stream'
+        self.files.append((fieldname, basename, mimetype, body))
+        return
+
+    def get_binary(self):
+        """Return a binary buffer containing the form data, including attached files."""
+        part_boundary = '--' + self.boundary
+
+        binary = io.BytesIO()
+        needsCLRF = False
```

Comment:
what's the point of having this `needs CLRF` flag when it doesn't appear to ever change?

Reply:
I does change, one first iteration there is no \r\n - but on subsequent loops there is.